# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/xul/nsXULTooltipListener.cpp
# Commit: 0a3da2090e48
# Full Hash: 0a3da2090e4859473f52b585bcb39baae2876b28
# Author: Jared Wein <jwein@mozilla.com>
# Date: 2021-02-26 03:55:47
# Regressor Bug: 1665390
# File Overlap Count: 1
# Description:
#   Bug 1665390 - Style bookmarks toolbar and tabs toolbar tooltips. r=NeilDeakin,ntim
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D104257
# ==============================================================================

diff -r 176b788753e8 -r 0a3da2090e48 layout/xul/nsXULTooltipListener.cpp
--- a/layout/xul/nsXULTooltipListener.cpp	Thu Feb 25 20:47:15 2021 +0000
+++ b/layout/xul/nsXULTooltipListener.cpp	Thu Feb 25 20:47:54 2021 +0000
@@ -458,8 +458,27 @@
   nsXULPopupManager* pm = nsXULPopupManager::GetInstance();
   if (pm) {
     nsCOMPtr<nsIContent> target = do_QueryReferent(mTargetNode);
-    pm->ShowTooltipAtScreen(currentTooltip, target, mMouseScreenX,
-                            mMouseScreenY);
+    nsAutoString position;
+    if (!target->AsElement()->Closest("treechildren"_ns, IgnoreErrors()) &&
+        !target->AsElement()->Closest("tree"_ns, IgnoreErrors())) {
+      nsAutoString closest;
+      currentTooltip->GetAttr(nsGkAtoms::anchortoclosest, closest);
+      if (!closest.IsEmpty()) {
+        target = target->AsElement()->Closest(NS_ConvertUTF16toUTF8(closest),
+                                              IgnoreErrors());
+        if (!target) {
+          target = do_QueryReferent(mTargetNode);
+        }
+      }
+      currentTooltip->GetAttr(nsGkAtoms::position, position);
+    }
+
+    if (position.IsEmpty()) {
+      pm->ShowTooltipAtScreen(currentTooltip, target, mMouseScreenX,
+                              mMouseScreenY);
+    } else {
+      pm->ShowTooltipAtPosition(currentTooltip, target, position);
+    }
 
     // Clear the current tooltip if the popup was not opened successfully.
     if (!pm->IsPopupOpen(currentTooltip)) mCurrentTooltip = nullptr;